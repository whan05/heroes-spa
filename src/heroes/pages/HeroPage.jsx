import { useMemo } from "react"
import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHreroById } from "../helpers"


export const HeroPage = () => {

  const {heroId} = useParams()

  const hero = useMemo(() => getHreroById(heroId), [heroId]) 

  const navigate =  useNavigate()

  const onNavigateBack = () => {
    navigate(-1)
  }

  if (!hero) {
    return <Navigate to="/marvel"/>
  }


  return (
    <div className="row mt-5">
      <div className="col-4">
        <img 
          src={`/src/assets/heroes/${hero.id}.jpg`}
          alt={hero.superhero} 
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>
      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item"><b>Alter ego:</b> {hero.alter_ego} </li>
          <li className="list-group-item"><b>Publisher:</b> {hero.publisher} </li>
          <li className="list-group-item"><b>First appearance:</b> {hero.first_appearance} </li>
        </ul>
        <h5 className="mt-3">Characters</h5>
        <p>{hero.characters}</p>
        <button 
          className="btn btn-outline-primary"
          onClick={onNavigateBack}
        >
          Back
        </button>
      </div>
    </div>
  )
}
