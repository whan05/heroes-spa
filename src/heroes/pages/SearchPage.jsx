import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import {HeroCard} from '../components';
import { getHeroesByName } from '../helpers';


export const SearchPage = () => {

  const navigate = useNavigate()
  const location = useLocation()

  const {q = ``} = queryString.parse( location.search )

  const heroes = getHeroesByName(q)

  const { searchText, onInputChange } = useForm({
    searchText: q
  })

  const onSearchSubmit = (e) => {
    e.preventDefault()

    // if (searchText.trim().length <=1 ) {
    //   return;
    // }

    navigate(`?q=${searchText}`)

  }

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit} aria-label="form">
            <input
              type="text"
              placeholder="Search a Hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
            <button
              className="btn btn-outline-primary mt-2"
            >
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {
            (q === "")
            ? <div className="alert alert-primary">
                Search a Hero
              </div>
            : (heroes.length === 0) 
            && 
              <div aria-label="alert-danger" className="alert alert-danger">
              No Hero with <b>{q}</b>
              </div>
          }
          
        </div>
      </div>
      {heroes.map( hero => (
        <HeroCard key={hero.id} {...hero}/>
      ))}

    </>
  )
}
