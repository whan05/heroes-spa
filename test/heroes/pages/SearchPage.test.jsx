/** @jest-environment jsdom */

import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn()

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockedUseNavigate
}))

describe('Pruebas en <Search Page />', () => {

    beforeEach(() => jest.clearAllMocks());
    
    test('debe de mostrarse correctamente con valores por defecto', () => {
        const {container} = render(
            <MemoryRouter>
                <SearchPage/>
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot()


    });

    test('debe mostrar a Batman y el input con el valor de queryString', () => {

        render(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <SearchPage/>
            </MemoryRouter>
        )
        const input = screen.getByRole("textbox")

        expect( input.value ).toBe("batman")
        const img = screen.getByRole("img")
        expect(img.src).toContain(`/src/assets/heroes/dc-batman.jpg`)

    });

    test('debe mostrar un error si no se encuentra el hero', () => {

        render(
            <MemoryRouter initialEntries={["/search?q=batman123"]}>
                <SearchPage/>
            </MemoryRouter>
        )
        
        const alert = screen.getByLabelText("alert-danger")
        expect(alert.textContent).toBe("No Hero with batman123")
        
    });

    test('debe de llamar el navigate a la pantalla nueva', () => {
        
        render(
            <MemoryRouter initialEntries={["/search"]}>
                <SearchPage/>
            </MemoryRouter>
        )

        const input = screen.getByRole("textbox")
        fireEvent.change( input, {target: {name: "searchText", value : "superman"}})
        
        const form =  screen.getByLabelText("form")
        fireEvent.submit(form)

        expect( mockedUseNavigate ).toHaveBeenCalled()
        expect( mockedUseNavigate ).toHaveBeenCalledWith("?q=superman")


    });

});