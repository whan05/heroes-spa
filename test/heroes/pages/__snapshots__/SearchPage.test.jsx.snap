// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Search Page /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a Hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;
